src/plaf/backends/halo2.rs

```diff
@@ -22,6 +22,44 @@ pub struct PlafH2Circuit {
     pub wit: Witness,
 }
 
+impl PlafH2Circuit {
+    pub fn instance<F: PrimeField<Repr = [u8; 32]>>(&self) -> Vec<Vec<F>> {
+        if !self.plaf.columns.public.is_empty() {
+            let mut instance_with_offsets = Vec::new();
+            for copy in &self.plaf.copys {
+                let (left_col, right_col) = &copy.columns;
+
+                let (witness_col, offsets) = match (left_col.kind, right_col.kind) {
+                    (ColumnKind::Witness, ColumnKind::Public) => (left_col, copy.offsets.clone()),
+                    (ColumnKind::Public, ColumnKind::Witness) => (
+                        right_col,
+                        copy.offsets.iter().map(|(l, r)| (*r, *l)).collect(),
+                    ),
+                    _ => continue,
+                };
+
+                for (witness_offset, public_offset) in offsets {
+                    instance_with_offsets.push((
+                        self.wit.witness[witness_col.index][witness_offset]
+                            .as_ref()
+                            .unwrap(),
+                        public_offset,
+                    ));
+                }
+            }
+            instance_with_offsets.sort_by_key(|&(_, offset)| offset);
+            let instance_values = instance_with_offsets
+                .iter()
+                .map(|&(val, _)| val.to_field())
+                .collect();
+
+            return vec![instance_values];
```

Comment:
It seems this implementation only works when there's a single public input column.  If the circuit uses `n` public input columns, the instance vector should contain `n` sub-vectors.  Nevertheless I think having multiple public input columns is not very common.
Could you add a check to panic whenever there are multiple public columns?  Something like `if self.plaf.columns.public.len() > 1 { unimplemented!("...") }`

Reply:
Didn't know that it was for multiple public columns haha. Thanks for explaining that :)